data-pipeline-flink/assessment-aggregator/src/main/scala/org/sunbird/dp/assessment/functions/UserScoreAggregateFunction.scala

```diff
@@ -17,67 +19,89 @@ import org.sunbird.dp.core.util.CassandraUtil
-          .where(QueryBuilder.eq("course_id", event.courseId)).and(QueryBuilder.eq("batch_id", event.batchId))
-          .and(QueryBuilder.eq("user_id", event.userId)).groupBy("user_id", "course_id", "batch_id", "content_id")
-        val rows: java.util.List[Row] = cassandraUtil.find(query.toString);
-        if (null != rows && !rows.isEmpty) {
-            rows.asScala.toList.map(row => {
-                "score:" + row.getString("content_id") -> row.getDouble("score").toInt
-            }).toMap
-        } else Map[String, Int]()
-    }
+  def getAggregateDetails(assessAggRows: List[Row]): List[String] = {
+    if (null != assessAggRows && !assessAggRows.isEmpty) {
+      assessAggRows.map { row =>
+        val aggMap = AggDetails(row.getString("attempt_id"), row.getTimestamp("last_attempted_on"),
+          row.getDouble("score"), row.getString("content_id"), row.getDouble("total_max_score"), config.aggType)
+        new Gson().toJson(aggMap)
+      }.toList
+    } else List()
+  }
 
-    def updateUserActivity(event: Event, score: Map[String, Int]): Unit = {
-        val scoreLastUpdatedTime: Map[String, Long] = score.map(m => m._1 -> System.currentTimeMillis())
-        val updateQuery = QueryBuilder.update(config.dbKeyspace, config.activityTable)
-          .`with`(QueryBuilder.putAll(config.agg, score.asJava))
-          .and(QueryBuilder.putAll(config.aggLastUpdated, scoreLastUpdatedTime.asJava))
-          .where(QueryBuilder.eq(config.activityId, event.courseId))
-          .and(QueryBuilder.eq(config.activityType, "Course"))
-          .and(QueryBuilder.eq(config.contextId, "cb:" + event.batchId))
-          .and(QueryBuilder.eq(config.activityUser, event.userId))
-        cassandraUtil.upsert(updateQuery.toString)
-        logger.info("Successfully updated scores in user activity  - batchid: "
-          + event.batchId + " ,userid: " + event.userId + " ,couserid: "
-          + event.courseId)
-    }
+  def getBestScore(event: Event): UserActivityAgg = {
+    val query = QueryBuilder.select().column("content_id").column("attempt_id").column("last_attempted_on").column("total_max_score").column("total_score").as("score").from(config.dbKeyspace, config.dbTable)
+      .where(QueryBuilder.eq("course_id", event.courseId)).and(QueryBuilder.eq("batch_id", event.batchId))
+      .and(QueryBuilder.eq("user_id", event.userId))
+    val rows = cassandraUtil.find(query.toString).asScala.toList
+    val aggDetailsList = getAggregateDetails(rows)
+
+    val aggMap = if (null != rows && !rows.isEmpty) {
+      val attemptGroupList = rows.groupBy(row => row.getString("content_id")).values
+      val topScores = attemptGroupList.map(row => row.maxBy(r => r.getDouble("score")))
+
+      topScores.flatMap(row => {
+        Map(s"score:${row.getString("content_id")}" -> row.getDouble("score"),
+          s"max_score:${row.getString("content_id")}" -> row.getDouble("total_max_score"))
+      }).toMap
+    } else Map[String, Double]()
```

Comment:
Similar to `getAggregateDetails` function, can you please put this in a function and call.